python/generate_compile_commands.py

```diff
@@ -0,0 +1,102 @@
+import os
+import json
+
+
+def get_filelist(root, whitelist):
+    """
+    1. Generates a file list dictionary of the structure:
+
+    {
+        pcl_folder: {
+            sub_folder: {
+                "impls": [list_of_impl_files]
+                "headers": [list_of_header_files]
+            }
+            ...
+        }
+        ...
+    }
+
+    """
+    pcl = {}
+
+    for folder in whitelist:
+        folder_path = os.path.join(root, folder)
+        if os.path.exists(folder_path):
+            impl_files = []
+            header_files = []
+
+            # walk the folder to get filenames
+            for root, _, files in os.walk(folder_path):
+                for file in files:
+                    # filter for implementation files
+                    if file.split(".")[-1] == "hpp":
+                        impl_files.append(os.path.join(root, file))
+                    # filter for header files
+                    if file.split(".")[-1] == "h":
+                        header_files.append(os.path.join(root, file))
+                    # add a filter for any other file type if desired
+                    _
+
+            # add list to pcl dict
+            pcl[folder] = {"impls": impl_files, "headers": header_files}
+
+        else:
+            raise Exception(f"{folder_path} doesn't exist")
+
+    return pcl
+
+
+def generate_compile_commmands(
+    file_list, directory_name, compiler_path, compiler_arguments
+):
+    """
+    2. Generating compilation commands database from the dictionary:
+
+    [
+        {
+            "directory": "build_directory",
+            "command": "path_to_compiler [compiler_argument_1, compiler_argument_2, ...] filename",
+            "file": "filename",
+        }
+        ...
+    ]
+
+    """
+    compile_commands = []
+
+    for contents in file_list.values():
+        for files in contents.values():
+            for file in files:
```

Comment:
A list comprehension might be more concise since we only really care about the inner loop anyways (no logic in the two outer loops), but I'm rusty on the syntax. Should be something akin to `for file in [files for files in contents.values() for contents in file_list.values()]:` iirc.

Reply:
The syntax is similar to "flattening"  the three lines above
```python
[ dict(directory=directory_name, command=f"{compiler_path} {compiler_arguments} {file}", file=file) for contents in file_list.values() for files in contents.values() for file in files]
```

Reply:
That'd be the complete method, what I meant to do to simplify but still keep readable was to keep only the innermost loop, by looping on a list comprehension and keeping the same body. Otherwise to me it seems a bit too much.

Reply:
committed @SergioRAgostinho's suggestion, let me know if you want me to use list comprehension instead.